### PR TITLE
Feature#139/pre battle phase

### LINKED ISSUE #139: pre-battle phase
- [x] 24h timer before time freezes
- [x] Freezing of time
- [x] Stopping continuation of actions during frozen time

### PR DESCRIPTION
**Pull request by Luktronic**
Implemented functionality regarding freezing of time, such as starting 24 hour timers when battles are declared or the stopping of movements and healings while time is frozen.

Additions:
- `TimeFreezeService` class which contains all the functionality for freezing of time, such as:
    - `start24hTimer()`  method which starts a 24 hour timer on a new virtual thread. As soon as the timer is done, a callback function will be called.
    - `freezeTime()` method, which sets the `isTimeFrozen` flag to `true`
- When battles are declared, a new 24h timer is started which freezes time on completion
- Ongoing movements and healings are now delayed and will not continue while time is frozen. The amount the action is delayed is the duration between the last delay/update and now. 

Closes #139

### CODE REVIEW

File: src/main/java/com/ardaslegends/domain/war/Battle.java [hunk 1]
```diff
@@ -59,9 +65,21 @@ public Battle(Set<War> wars, String name, Set<Army> attackingArmies, Set<Army> d
...
+
+    public Army getInitialAttacker() {
+        return attackingArmies.stream().findFirst()
+                .orElseThrow(() -> new NullPointerException("Found no initial attacking army in battle at location %s".formatted(battleLocation.toString())));
+    }
+
+    public Army getFirstDefender() {
+        return defendingArmies.stream().findFirst()
+                .orElseThrow(() -> new NullPointerException("Found no defending armies in battle at location %s".formatted(battleLocation.toString())));
+    }
```
Comment (GroellerKarim) [inline src/main/java/com/ardaslegends/domain/war/Battle.java:83]: is a set ordered?

**Review by GroellerKarim — APPROVED** [review]

File: src/main/java/com/ardaslegends/domain/war/Battle.java (hunk 1 above)
Comment (Luktronic) [reply src/main/java/com/ardaslegends/domain/war/Battle.java:83]: No, which is why I implemented another solution for this on the new branch.

You did the same thing in War.java btw :Pepelaugh:

Comment (Luktronic) [reply src/main/java/com/ardaslegends/domain/war/Battle.java:83]: Actually no, you did it differently

Comment (Luktronic) [reply src/main/java/com/ardaslegends/domain/war/Battle.java:83]: Just remembered

Comment (GroellerKarim) [reply src/main/java/com/ardaslegends/domain/war/Battle.java:83]: xD

**Pull request merged**